apps/ejabberd/src/ejabberd_router.erl

```diff
@@ -118,51 +124,133 @@ route_error(From, To, ErrPacket, OrigPacket) ->
 
 -spec register_components([Domain :: domain()]) -> ok | {error, any()}.
 register_components(Domains) ->
+    register_components(Domains, node()).
+
+-spec register_components([Domain :: domain()], Node :: node()) -> ok | {error, any()}.
+register_components(Domains, Node) ->
     LDomains = [{jid:nameprep(Domain), Domain} || Domain <- Domains],
     Handler = make_handler(undefined),
     F = fun() ->
-            [do_register_component(LDomain, Handler) || LDomain <- LDomains],
+            [do_register_component(LDomain, Handler, Node) || LDomain <- LDomains],
             ok
     end,
     case mnesia:transaction(F) of
         {atomic, ok}      -> ok;
         {aborted, Reason} -> {error, Reason}
     end.
 
+%% @doc
+%% components are registered in two places: external_components table as local components
+%% and external_components_global as globals. Registration should be done by register_component/1
+%% or register_components/1, which registers them for current node; the arity 2 funcs are
+%% here for testing.
 -spec register_component(Domain :: domain()) -> ok | {error, any()}.
 register_component(Domain) ->
-    register_components([Domain]).
+    register_components([Domain], node()).
+
+-spec register_component(Domain :: domain(), Node :: node()) -> ok | {error, any()}.
+register_component(Domain, Node) ->
+    register_components([Domain], Node).
 
-do_register_component({error, Domain}, _Handler) ->
+do_register_component({error, Domain}, _Handler, _Node) ->
     error({invalid_domain, Domain});
-do_register_component({LDomain, _}, Handler) ->
-    Component = #external_component{domain = LDomain, handler = Handler},
-    case {mnesia:read(route, LDomain),
-          mnesia:read(external_component, LDomain)} of
-        {[], []} ->
-            ok = mnesia:write(Component);
+do_register_component({LDomain, _}, Handler, Node) ->
+    case check_component(LDomain, Node) of
+        ok ->
+            ComponentGlobal = #external_component{domain = LDomain, handler = Handler, node = Node},
+            mnesia:write(external_component_global, ComponentGlobal, write),
+            NDomain = {LDomain, Node},
+            Component = #external_component{domain = NDomain, handler = Handler, node = Node},
+            mnesia:write(Component);
+        _ -> mnesia:abort(route_already_exists)
+    end.
+
+check_component(LDomain, Node) ->
+    check_component(route, LDomain, Node).
+
+check_component(route, LDomain, Node) ->
```

Comment:
IMO, it would be more readable to have 3 differently named function instead of pattern matching since the atoms `external_comp_local` and `external_com_global` are used only for this.